### DISCUSSION TITLE
collision check in  manipulators

### DISCUSSION THREAD

**Discussion by leiyu2023** [tesseract-robotics/tesseract#939] [Q&A] [answered]
Hello, I have two questions about collision detection that are deeply troubling me：
1.I need to calculate the minimum distance of all links in my application, the CollisionMarginData will greatly affect the collision checking time, the smaller the valve execution time, I checked the document and did not find a description about it, can you briefly explain the principle behind it?If I set a small threshold will I miss some collision pairs?
2.The collision volume of the link in my urdf is concave and defined by an stl file, so I used vhacd decomposition to get a new stl file to replace the original one, but I found that the collision checking algorithm's execution time is slower than the original one.
Am I using it in a wrong way? What will tesseract do if I import the original concave body directly?

Can you give me some advice?@Levi-Armstrong

**Comment by Levi-Armstrong** [comment, marked as answer]
1. The way most collision checkers works it they are broken up into two phases. The first is called the broad phase and the second is called the narrow phase. The narrow phase is very expensive step which will find the distance, normal, manifold, etc., so in most application you do not want to do this unless the two shapes are within some threshold which leads me to the broad phase. The broad phase consists of check each collision objects bounding volume which is usually an axis aligned bounding box but there several different methods for this. The process of checking each shapes bounding volume box to every other box and then only perform the narrow phase if the bounding volume box intersect. This make collision checking much faster except in the use case that you are describing because you want to run narrow phase only because you need the distance regardless so the broad phase is unnecessary. So in your case you need to set the collision margin high so you get the data you desire. Though as you pointed out this will increase collision checking time significantly. I am not sure if the underlying collision checkers support disabling the broad phase check but I would be in favor of adding this you are interested in adding this feature.
2. This is expected because by running convex decomposition you have increased the number of collision shapes in the environment and for each collision object pair it has to the run the narrow phase. Another option would be to see if you could represent the collision object using primitive shapes. Also by default tesseract will convert all provided collision geometries to convex hulls under the hood.

**Reply by Levi-Armstrong** [reply]
Oh, also for your use case the Bullet Simple Collision check may perform better. You can set it as the default in the collision checker in the yaml file to see if it performs better.

**Reply by leiyu2023** [reply]
Thanks for your clear answer, it solved a lot of my doubts. As you said BulletDiscreteSimpleManager is faster than FCLDiscreteBVHManager. I will check if fcl supports disabling the broad phase check.  I still have some questions: 
1.Why is the speed of BulletDiscreteSimpleManager also affected by CollisionMarginData? 
2.Why each collision object pair it has to the run the narrow phase after using convex decomposition？
3.How exactly does tesseract differentiate between stl files that have been processed by vhacd and those that have not?
Also, I have a question about practice: 
is there a way to specify collision pairs directly instead of defaulting to all considered collisions and then disabling collisions pairs via the srdf file.

**Comment by Levi-Armstrong** [comment]
> Why is the speed of BulletDiscreteSimpleManager also affected by CollisionMarginData?

While it does not use the broadphase component it does still do a bounding volume check. I think an option could be added to this and possibly all of the collision checkers to avoid bounding volumen check and go straight to the narrow phase. 

> Why each collision object pair it has to the run the narrow phase after using convex decomposition？

The narrow phase is the algorithm which calculates the distance between the two shapes so it must be ran if you want distance. This algorithm is different based  on the shapes types. If they are both convex hulls it runs the GJK algorithm and if primitive shapes they have custom math which is much faster. 

> How exactly does tesseract differentiate between stl files that have been processed by vhacd and those that have not?

The default behavior converts everything to convex hulls. This can be overriden by setting the urdf version to 2 and then you have to explicitly use convex_hull and mesh types in the urdf.

> is there a way to specify collision pairs directly instead of defaulting to all considered collisions and then disabling collisions pairs via the srdf file.

Not at the moment. Though by default it does not check static objects to other static objects, it only checks active->active and active->static so you need to make sure you are setting the active links when setting up the collision checker. This can also have a significant impact on performance if you are not setting the active links.

**Reply by leiyu2023** [reply]
Thanks for the detailed answer . I have tried the effect of setting the active links in the checker. The execution is 10 times faster after setting it, but I also noticed that the returned results have a lot fewer collision pairs than before when it was not set. I examined the code in detail and determined that I had set all links to active. the code is below:

```
checker=t_env.getDiscreteContactManager("BulletDiscreteSimpleManager")
checker.setCollisionMarginData(tesseract_common.CollisionMarginData(10))
checker.setActiveCollisionObjects(t_env.getJointNames())

```

**Reply by Levi-Armstrong** [reply]
Do you need distance between objects that do not move?

**Reply by leiyu2023** [reply]
Thanks for your reply, in fact, I only considered the self-collision distance, and among all collision objects only base is not able to move, so I set all links to be active, and just now I found some mistakes in the code: I should use t_env.getActiveLinkNames() instead of t_env. getJointNames().
